language/move-lang/src/shared/unique_map.rs

```diff
@@ -25,49 +25,50 @@ impl<K: TName, V> UniqueMap<K, V> {
         self.0.len()
     }
 
-    pub fn add(&mut self, key: K, value: V) -> Result<(), K::Loc> {
+    pub fn add(&mut self, key: K, value: V) -> Result<(), (K, K::Loc)> {
+        if let Some(old_loc) = self.get_loc(&key) {
+            return Err((key, *old_loc));
+        }
         let (loc, key_) = key.drop_loc();
         let old_value = self.0.insert(key_, (loc, value));
-        if let Some((old_loc, _)) = old_value {
-            return Err(old_loc);
-        }
```

Comment:
This should be the reason why the unit tests are failing with the following pattern:

```diff
   ┌── tests/move_check/typing/bind_duplicate_binding.move:6:23 ───
   │
 6 │         let (f, R{f}, f) = (0, R { f: 0 }, 0);
   │                       ^ Duplicate declaration for local 'f' in a given 'let'
   ·
 6 │         let (f, R{f}, f) = (0, R { f: 0 }, 0);
-  │                   - Previously declared here
+  │              - Previously declared here
   │
```

It is a matter of placing the `Previously declared here` at the first seen location or the last seen location :)

To me it seems to be a harmless one though.

Besides, test case `translated_ir_tests/move/generics/instantiation_loops/recursive_struct.move` needs an update? The last `module M3` needs to be renamed to `M4`.

Reply:
Good catch and detective work! Saved me having to look this morning :) 

Updated and fixed